servant-server/test/Servant/Server/UsingConfigSpec.hs

```diff
@@ -0,0 +1,105 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Servant.Server.UsingConfigSpec where
+
+import           Control.Monad.Trans.Except
+import           Network.Wai
+import           Test.Hspec (Spec, describe, it)
+import           Test.Hspec.Wai
+
+import           Servant
+import           Servant.Server.UsingConfigSpec.TestCombinators
+
+-- * API
+
+type OneEntryAPI =
+  ExtractFromConfig :> Get '[JSON] String
+
+testServer :: String -> ExceptT ServantErr IO String
+testServer s = return s
+
+oneEntryApp :: Application
+oneEntryApp =
+  serve (Proxy :: Proxy OneEntryAPI) config testServer
+  where
+    config :: Config '[String]
+    config = "configEntry" :. EmptyConfig
+
+type OneEntryTwiceAPI =
+  "foo" :> ExtractFromConfig :> Get '[JSON] String :<|>
+  "bar" :> ExtractFromConfig :> Get '[JSON] String
+
+oneEntryTwiceApp :: Application
+oneEntryTwiceApp = serve (Proxy :: Proxy OneEntryTwiceAPI) config $
+  testServer :<|>
+  testServer
+  where
+    config :: Config '[String]
+    config = "configEntryTwice" :. EmptyConfig
+
+-- * tests
+
+spec :: Spec
+spec = do
+  describe "accessing config entries from custom combinators" $ do
+    with (return oneEntryApp) $ do
+      it "allows to retrieve a ConfigEntry" $ do
+        get "/" `shouldRespondWith` "\"configEntry\""
+
+    with (return oneEntryTwiceApp) $ do
+      it "allows to retrieve the same ConfigEntry twice" $ do
+        get "/foo" `shouldRespondWith` "\"configEntryTwice\""
+        get "/bar" `shouldRespondWith` "\"configEntryTwice\""
+
+  spec2
+
+type InjectAPI =
+  InjectIntoConfig :> "untagged" :> ExtractFromConfig :>
+    Get '[JSON] String :<|>
+  InjectIntoConfig :> "tagged" :> ExtractFromConfig :>
+    Get '[JSON] String
+
+injectApp :: Application
+injectApp = serve (Proxy :: Proxy InjectAPI) config $
+  (\ s -> return s) :<|>
+  (\ s -> return ("tagged: " ++ s))
+  where
+    config = EmptyConfig
+
+spec2 :: Spec
+spec2 = do
```

Comment:
Any reason to have each spec call the next one, rather than have a top-level spec that calls all of them (as we do elsewhere)? Also, the names are a bit generic...